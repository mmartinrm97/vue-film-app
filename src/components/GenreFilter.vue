<template>
  <div class="p-4 bg-gray-200 dark:bg-gray-900 text-gray-800 dark:text-gray-300 rounded dark:shadow-lg shadow-lg dark:overflow-hidden overflow-hidden">

    <!-- Display  name filter-->
    <div class="mb-4">
      <label for="name" class="block text-gray-700 dark:text-gray-300 text-sm font-bold mb-2"
        >Name:</label
      >
      <input
        id="name"
        v-model.trim="nameFilter"
        type="text"
        placeholder="Enter movie name"
        class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 dark:text-gray-300 leading-tight focus:outline-none focus:shadow-outline dark:focus:shadow-outline"
      />
    </div>

    <!-- Display description filter -->
    <div class="mb-4">
      <label for="description" class="block text-gray-700 dark:text-gray-300 text-sm font-bold mb-2"
        >Description:</label
      >
      <input
        id="description"
        v-model.trim="descriptionFilter"
        type="text"
        placeholder="Enter movie description"
        class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 dark:text-gray-300 leading-tight focus:outline-none focus:shadow-outline dark:focus:shadow-outline"
      />
    </div>
        <!-- Display genre filter -->
        <div class="mb-4">
      <label for="genre" class="block text-gray-700 dark:text-gray-300 text-sm font-bold mb-2"
        >Genre:</label
      >
      <input
        id="genre"
        v-model.trim="selectedGenre"
        @change="addGenreFilter"
        class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 dark:text-gray-300 leading-tight focus:outline-none focus:shadow-outline dark:focus:shadow-outline"
        list="genre-list"
        placeholder="Enter movie genre"
      />
      <datalist id="genre-list" class="w-full">
        <!-- Define la lista de géneros -->
        <option v-for="genre in genresApiData" :key="genre.id" :value="genre.name">
          {{ genre.name }}
        </option>
      </datalist>
      <span v-if="!isValidGenre" class="text-red-600 text-sm"
        >"{{ selectedGenre }}" is not a valid genre</span
      >
    </div>

    <!-- Display selected genres as labels -->
    <div class="mb-2">
      <span
        v-for="genre in genresFilter"
        :key="genre"
        class="inline-block bg-gray-300 text-gray-800 text-xs font-semibold rounded-full px-2 py-1 mr-1"
      >
        {{ genre }}
        <button @click="removeGenreFilter(genre)" class="ml-1 text-gray-600 text-xs">×</button>
      </span>
    </div>
  </div>
</template>

<script setup lang="ts">
import filmsApi from '@/api/movies.api'
import type { MovieFilter, Genre } from '@/movies/types/movie'
import { ref, onMounted, watch } from 'vue'

const emit = defineEmits<{
  (event: 'updateFilters', filters: MovieFilter): void
}>()

// Data from API
const genresApiData = ref<Genre[]>([])

// Selected genre
const selectedGenre = ref<string>('')

// Constant to validate genre 
//TODO: Validate genre
const isValidGenre = ref(true)

// Filters to send to the MovieList component
const genresFilter = ref<string[]>([])
const nameFilter = ref<string>('')
const descriptionFilter = ref<string>('')

watch([selectedGenre, nameFilter, descriptionFilter], () => {
  const newFilters: MovieFilter = {
    genre: genresFilter.value,
    name: nameFilter.value,
    description: descriptionFilter.value
  }
  emit('updateFilters', newFilters)
})

function addGenreFilter() {
  if (selectedGenre.value !== '' && !genresFilter.value.includes(selectedGenre.value)) {
    genresFilter.value.push(selectedGenre.value)
  }
  selectedGenre.value = ''
}

function removeGenreFilter(genre: string) {
  const index = genresFilter.value.indexOf(genre)
  if (index !== -1) {
    genresFilter.value.splice(index, 1)
  }
}

onMounted(async () => {
  const result = await filmsApi.get<Genre[]>('/genres')
  genresApiData.value = result.data
})
</script>

<style scoped></style>
