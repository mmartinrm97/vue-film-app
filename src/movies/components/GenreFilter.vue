<template>
  <div
    class="flex flex-col gap-y-6 overflow-hidden rounded-lg bg-sky-50 p-4 text-gray-800 shadow-lg dark:overflow-hidden dark:bg-sky-950 dark:text-gray-300"
  >
    <div>
      <!-- Filter tile -->
      <h2>
        <span class="text-xl font-semibold text-gray-900 dark:text-gray-100">Filters:</span>
        <span class="ml-2 text-sm text-gray-700 dark:text-gray-300">({{ totalFilters }})</span>
      </h2>
    </div>
    <!-- Display  name filter-->
    <div>
      <label for="name" class="mb-2 block text-sm font-bold text-gray-700 dark:text-gray-300"
        >Name:</label
      >
      <input
        :disabled="hasError"
        id="name"
        v-model.trim="nameFilter"
        type="text"
        placeholder="Enter movie name"
        class="focus:shadow-outline dark:focus:shadow-outline w-full appearance-none rounded border px-3 py-2 leading-tight text-gray-700 shadow focus:outline-none disabled:cursor-not-allowed"
      />
    </div>

    <!-- Display description filter -->
    <div>
      <label for="description" class="mb-2 block text-sm font-bold text-gray-700 dark:text-gray-300"
        >Description:</label
      >
      <textarea
        :disabled="hasError"
        id="description"
        v-model.trim="descriptionFilter"
        type="texta"
        placeholder="Enter movie description"
        class="focus:shadow-outline dark:focus:shadow-outline w-full appearance-none rounded border px-3 py-2 leading-tight text-gray-700 shadow focus:outline-none disabled:cursor-not-allowed"
      />
    </div>
    <!-- Display genre filter -->
    <div>
      <label for="genre" class="mb-2 block text-sm font-bold text-gray-700 dark:text-gray-300"
        >Genre:</label
      >
      <input
        id="genre"
        :disabled="hasError"
        v-model.trim="selectedGenre"
        @change="addGenreFilter"
        class="focus:shadow-outline dark:focus:shadow-outline w-full appearance-none rounded border px-3 py-2 leading-tight text-gray-700 shadow focus:outline-none disabled:cursor-not-allowed"
        list="genre-list"
        placeholder="Enter movie genre"
      />
      <datalist id="genre-list" class="w-full">
        <!-- Define la lista de géneros -->
        <option v-for="genre in genresApiData" :key="genre.id" :value="genre.name">
          {{ genre.name }}
        </option>
      </datalist>
      <span v-if="!isValidGenre" class="text-sm text-red-600"
        >"{{ selectedGenre }}" is not a valid genre</span
      >
    </div>

    <!-- Display selected genres as labels -->
    <div>
      <span
        v-for="genre in genresFilter"
        :key="genre"
        class="mr-1 inline-block rounded-full bg-gray-300 px-2 py-1 text-xs font-semibold text-gray-800"
      >
        {{ genre }}
        <button @click="removeGenreFilter(genre)" class="ml-1 text-xs text-gray-600">×</button>
      </span>
    </div>
  </div>
</template>

<script setup lang="ts">
import { ref, onMounted, watch, computed } from 'vue'
import type { MovieFilter, MovieGenre } from '@/movies/types'
import { useGetGenresApi } from '../composables/useGetGenresApi'

const emit = defineEmits<{
  (event: 'updateFilters', filters: MovieFilter): void
}>()

// Data from API
const genresApiData = ref<MovieGenre[]>([])

// Selected genre
const selectedGenre = ref<string>('')

// Constant to validate genre
//TODO: Validate genre
const isValidGenre = ref(true)

// Filters to send to the MovieList component
const genresFilter = ref<string[]>([])
const nameFilter = ref<string>('')
const descriptionFilter = ref<string>('')

// total Filters
const totalFilters = computed(() => {
  let count = 0

  if (nameFilter.value.trim() !== '') {
    count++
  }
  if (descriptionFilter.value.trim() !== '') {
    count++
  }
  count += genresFilter.value.length

  return count
})

watch([selectedGenre, nameFilter, descriptionFilter], () => {
  const newFilters: MovieFilter = {
    genre: genresFilter.value,
    name: nameFilter.value,
    description: descriptionFilter.value
  }
  emit('updateFilters', newFilters)
})

function addGenreFilter() {
  if (selectedGenre.value !== '' && !genresFilter.value.includes(selectedGenre.value)) {
    genresFilter.value.push(selectedGenre.value)
  }
  selectedGenre.value = ''
}

function removeGenreFilter(genre: string) {
  const index = genresFilter.value.indexOf(genre)
  if (index !== -1) {
    genresFilter.value.splice(index, 1)
  }
}

const hasError = ref<boolean>(true)

async function loadGenres() {
  hasError.value = true
  try {
    const data = await useGetGenresApi()
    genresApiData.value = data
    hasError.value = false
  } catch (error: any) {
    hasError.value = true
  }
}

onMounted(async () => {
  await loadGenres()
})
</script>

<style scoped></style>
