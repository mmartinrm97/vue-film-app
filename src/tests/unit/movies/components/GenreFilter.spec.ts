import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import GenreFilter from '@/movies/components/GenreFilter.vue'
import type { MovieFilter } from '@/movies/types'

describe('GenreFilter component', async () => {
  it('renders the component with initial state', () => {
    const filters: MovieFilter = {
      genre: [],
      name: '',
      description: ''
    }

    const wrapper = mount(GenreFilter, {
      props: { filters }
    })

    // Verifica que el componente se ha renderizado correctamente
    expect(wrapper.exists()).toBeTruthy

    //Genera un snapshot
    expect(wrapper.html()).toMatchSnapshot()

    // Verifica que el título de los filtros esté presente
    const title = wrapper.find('span')
    expect(title.text()).toBe('Filters:')

    // Verifica que el contador de filtros inicia en 0
    expect(wrapper.text()).toContain('(0)')

    // Verifica la presencia de elementos filtros HTML en el componente
    expect(wrapper.find('label[for="name"]').exists()).toBeTruthy
    expect(wrapper.find('input[id="name"]').exists()).toBeTruthy
    expect(wrapper.find('label[for="description"]').exists()).toBeTruthy
    expect(wrapper.find('textarea[id="description"]').exists()).toBeTruthy
    expect(wrapper.find('label[for="genre"]').exists()).toBeTruthy
    expect(wrapper.find('input[id="genre"]').exists()).toBeTruthy
  })

  it('disables inputs when hasError is true', async () => {
    const filters: MovieFilter = {
      genre: [],
      name: '',
      description: ''
    }

    const wrapper = mount(GenreFilter, {
      props: { filters }
    })

    // Verifica que los inputs estén habilitados por defecto
    const nameInput = wrapper.find('#name')
    const descriptionInput = wrapper.find('#description')
    const genreInput = wrapper.find('#genre')

    expect(nameInput.attributes('disabled')).toBeFalsy()
    expect(descriptionInput.attributes('disabled')).toBeFalsy()
    expect(genreInput.attributes('disabled')).toBeFalsy()
  })
})
