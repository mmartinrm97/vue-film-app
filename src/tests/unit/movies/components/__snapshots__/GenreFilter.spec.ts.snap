// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GenreFilter component > renders the component with initial state 1`] = `
"<div class=\\"flex flex-col gap-y-6 overflow-hidden rounded-lg bg-sky-50 p-4 text-gray-800 shadow-lg dark:overflow-hidden dark:bg-sky-950 dark:text-gray-300\\" filters=\\"[object Object]\\">
  <div>
    <!-- Filter tile -->
    <h2><span class=\\"text-xl font-semibold text-gray-900 dark:text-gray-100\\">Filters:</span><span class=\\"ml-2 text-sm text-gray-700 dark:text-gray-300\\">(0)</span></h2>
  </div><!-- Display  name filter-->
  <div><label for=\\"name\\" class=\\"mb-2 block text-sm font-bold text-gray-700 dark:text-gray-300\\">Name:</label><input disabled=\\"\\" id=\\"name\\" type=\\"text\\" placeholder=\\"Enter movie name\\" class=\\"focus:shadow-outline dark:focus:shadow-outline w-full appearance-none rounded border px-3 py-2 leading-tight text-gray-700 shadow focus:outline-none disabled:cursor-not-allowed\\"></div><!-- Display description filter -->
  <div><label for=\\"description\\" class=\\"mb-2 block text-sm font-bold text-gray-700 dark:text-gray-300\\">Description:</label><textarea disabled=\\"\\" id=\\"description\\" type=\\"texta\\" placeholder=\\"Enter movie description\\" class=\\"focus:shadow-outline dark:focus:shadow-outline w-full appearance-none rounded border px-3 py-2 leading-tight text-gray-700 shadow focus:outline-none disabled:cursor-not-allowed\\"></textarea></div><!-- Display genre filter -->
  <div><label for=\\"genre\\" class=\\"mb-2 block text-sm font-bold text-gray-700 dark:text-gray-300\\">Genre:</label><input id=\\"genre\\" disabled=\\"\\" class=\\"focus:shadow-outline dark:focus:shadow-outline w-full appearance-none rounded border px-3 py-2 leading-tight text-gray-700 shadow focus:outline-none disabled:cursor-not-allowed\\" list=\\"genre-list\\" placeholder=\\"Enter movie genre\\"><datalist id=\\"genre-list\\" class=\\"w-full\\">
      <!-- Define la lista de gÃ©neros -->
    </datalist>
    <!--v-if-->
  </div><!-- Display selected genres as labels -->
  <div></div>
</div>"
`;
